### PR TITLE
Add implementation for get data using tile key.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Get QuadTreeIndex for requested tile with debth equal 4, convert to partitions.
Put partitions to cache. Get blob data using data handle for requested tile key. Return data to user.

Relates-To: OLPEDGE-1529

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h [hunk 1]
```diff
@@ -29,10 +29,16 @@ namespace olp {
 namespace dataservice {
 namespace read {
 class DataRequest;
+class TileRequest;
 namespace repository {
 
 class DataRepository final {
  public:
+  static DataResponse GetVersionedDataTileQuadTree(
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h]:
```suggestion
  static DataResponse GetVersionedTile(
```
How we do it internally is our business. Doesn't need to be extra marked in API.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h [hunk 2]
```diff
@@ -29,10 +29,16 @@ namespace olp {
 namespace dataservice {
 namespace read {
 class DataRequest;
+class TileRequest;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h]: This doesn't work. Actually you can remove both fwd defines.
We already include Types.h and fwd defines only work for references or ptrs.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 3]
```diff
@@ -40,6 +41,7 @@ using namespace olp::client;
 
 namespace {
 constexpr auto kLogTag = "PartitionsRepository";
+constexpr std::uint32_t kMaxQuadTreeIndexDepth = 4u;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]:
```suggestion
constexpr uint32_t kMaxQuadTreeIndexDepth = 4u;
```

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h [hunk 4]
```diff
@@ -56,6 +58,21 @@ class PartitionsRepository final {
...
+      TileRequest request, int64_t version, client::CancellationContext context,
+      client::OlpClientSettings settings,
+      std::string& requested_tile_data_handle);
+
+  static model::Partitions GetTileFromCache(const client::HRN& catalog,
+                                            const std::string& layer_id,
+                                            TileRequest request,
+                                            int64_t version,
+                                            client::OlpClientSettings settings);
+
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: Not sure I get it.
I assumed we have something like `GetVersionedTile(...)` which looks in the cache if the tiles partition is there, if not goes to the `QueryApi::QuadTreeIndex(...)` with depth 4 which downloads quad-tree and caches each subquad as partition metadata same as `PrefetchTilesRepository::GetSubQuads()`. At the end it returns the requested partition to the user.
Additional improvements can be done later on, like:
- If requested tile key not found i cache check if parent is found and if yes then assume there was a quad tree index query before but it maybe got LRU evicted in the meantime so we maybe only downlaod that specific tile in the end.
-Return nearest tile if the tile is not in the response (optional, old hype behaviour)

Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: 1.Updated, please check. Added check if parent tile in cache. If true, get requested tile. Added test GetVersionedDataTileQuadTreeParentTileInCache.
2. Return nearest tile if the tile is not in the response. Is this applicable only for CacheOnly option?

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h (hunk 2 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h]: There is no DataRequest in Types.h. Forward declarations changed to includes.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h (hunk 1 above)
Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h]: My point is we don't need to add implementation details to the method name. Just call it `GetVersionedTile`. How we get the tile or the metadata internally is not important for the API name.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp [hunk 5]
```diff
@@ -49,6 +50,74 @@ constexpr auto kBlobService = "blob";
...
+          catalog, layer_id, data_request, context, settings);
+    }
+
+    auto response = PartitionsRepository::QueryPartitionsAndGetDataHandle(
+        catalog, layer_id, request, version, context, settings,
+        requested_tile_data_handle);
+    if (!response.IsSuccessful()) {
+      return response.GetError();
+    }
+  }
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: All this should not be here, it should be implementation details of the `PartitionsRepository::GetVersionedTile()`.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h (hunk 4 above)
Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: Those were only theoretical scenarios for the future. At least the 2. point I would omit for now as I don't see any requirement.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h [hunk 6]
```diff
@@ -56,6 +58,21 @@ class PartitionsRepository final {
...
+      const client::HRN& catalog, const std::string& layer_id,
+      TileRequest request, int64_t version, client::CancellationContext context,
+      client::OlpClientSettings settings,
+      std::string& requested_tile_data_handle);
+
+  static model::Partitions GetTileFromCache(const client::HRN& catalog,
+                                            const std::string& layer_id,
+                                            TileRequest request,
+                                            int64_t version,
+                                            client::OlpClientSettings settings);
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: Why not only `GetVersionedTile()`?
Why do you need three public methods if all three serve actually an internal implementation purpose and are not in any point related to what DataRepository needs?
DataRepository::GetVersionedTile() is only interested in the partition metadata for his request. How you get this partition metadata (by using quad tree index, by requesting single partition) is implementation details of PartitionsRepository and should not be handled by DataRepository.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp (hunk 5 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: Please check the other methods in this class, partition repo works with the metadata only.

Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: This is an implementation detail of PartitionsRepo which is easily fixed, considering that the query REST API wrapper is static and can be used everywhere.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h (hunk 6 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: I can move this methods to DataRepository and make them private. It was spitted to do not have very big method, and theoretically this methods could be useful for getting metadata for partitions in future. Also it easier to test and cover all functionality separately. Should I move everything to one function?

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h (hunk 1 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.h]: Renamed

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h (hunk 4 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.h]: Removed checking for parent tile in cache.

File: olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp (hunk 5 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp]: Moved to PartitionRepository

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 7]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+    TileRequest request, int64_t version, client::CancellationContext context,
+    client::OlpClientSettings settings) {
+  auto fetch_option = request.GetFetchOption();
+  auto tile = request.GetTileKey().ToHereTile();
+  auto query_api = ApiClientLookup::LookupApi(
+      catalog, context, "query", "v1", request.GetFetchOption(), settings);
+
+  if (!query_api.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR(kLogTag,
+                      "QueryPartitionsAndGetDataHandle: LookupApi failed.");
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]:
```suggestion
                      "QueryPartitionForVersionedTile: LookupApi failed.");
```

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 8]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+                        "QuadTreeIndex failed (%s, %" PRId64 ", %" PRId32 ")",
+                        tile.c_str(), version, kMaxQuadTreeIndexDepth);
+    return quad_tree.GetError();
+  }
+  model::Partitions result;
+  model::Partitions partitions_for_caching;
+  const auto& subquads = quad_tree.GetResult().GetSubQuads();
+  partitions_for_caching.GetMutablePartitions().reserve(subquads.size());
+
+  OLP_SDK_LOG_TRACE_F(kLogTag, "Requested tile subquads size %lu.",
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]:
```suggestion
  OLP_SDK_LOG_DEBUG_F(kLogTag, "Requested tile subquads size %lu.",
```

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 9]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+  if (!quad_tree.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "QuadTreeIndex failed (%s, %" PRId64 ", %" PRId32 ")",
+                        tile.c_str(), version, kMaxQuadTreeIndexDepth);
+    return quad_tree.GetError();
+  }
+  model::Partitions result;
+  model::Partitions partitions_for_caching;
+  const auto& subquads = quad_tree.GetResult().GetSubQuads();
+  partitions_for_caching.GetMutablePartitions().reserve(subquads.size());
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]: Please extract to a local variable partitions_for_caching.GetMutablePartitions()

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 10]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+
+  OLP_SDK_LOG_TRACE_F(kLogTag, "Requested tile subquads size %lu.",
+                      subquads.size());
+
+  for (const auto& subquad : subquads) {
+    auto subtile =
+        request.GetTileKey().AddedSubHereTile(subquad->GetSubQuadKey());
+    // add partitions for caching
+    partitions_for_caching.GetMutablePartitions().emplace_back(
+        PartitionsRepository::PartitionFromSubQuad(*subquad,
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]:
```suggestion
       PartitionFromSubQuad(*subquad, subtile.ToHereTile()));
```

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 11]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+                                                   subtile.ToHereTile()));
+    // find data handle for requested tile
+    if (subtile.ToHereTile().compare(tile) == 0) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]: You compare tiles as string, but you can compare them as TileKeys. Why?

`subtile.ToHereTile().compare(tile.ToHereTile())` is same as `subtile == tile`.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 12]
```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
...
+      boost::none, request.GetBillingTag(), context);
+
+  if (!quad_tree.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "QuadTreeIndex failed (%s, %" PRId64 ", %" PRId32 ")",
+                        tile.c_str(), version, kMaxQuadTreeIndexDepth);
+    return quad_tree.GetError();
+  }
+  model::Partitions result;
+  model::Partitions partitions_for_caching;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp]:
```suggestion
  model::Partitions partitions;
```
Would simplify a lot the formatting, I presume.

**Pull request merged**